File: resolveai/agent/explain.py
"""Operator-facing explanations: WHAT the agent did, WHY, on WHAT evidence, and WHAT happens next.

Built deterministically from the typed state (decision, escalation rule, output-gate checks, verification, handoff and
clarification packets). No LLM and no free-text reasoning: every sentence is assembled from recorded fields, so it can
be audited against the trace.
"""
from __future__ import annotations

from resolveai.agent.state import AgentState
from resolveai.schemas.core import DecisionSummary, EvidenceRef
from resolveai.schemas.evidence import EvidenceSet


def citations(st: AgentState) -> list[EvidenceRef]:
    """The historical cases an autonomous troubleshooting reply was drafted from, with provenance and rank metadata."""
    if st.draft is None or st.evidence is None:
        return []
    by_id = {i.evidence_id: i for i in st.evidence.items}
    refs = []
    for eid in st.draft.evidence_ids:
        it = by_id.get(eid)
        if it is None:
            continue
        refs.append(EvidenceRef(evidence_id=it.evidence_id, thread_id=it.thread_id, source_row_id=it.source_row_id, created_at=it.created_at, rank=it.rank,
                                similarity=round(it.quality.semantic_relevance, 3), retrieval_source=it.retrieval_source, action_class=it.quality.action_class,
                                resolution_bearing=it.quality.resolution_relevance, outcome=it.outcome))
    return refs


def evidence_line(ev: EvidenceSet | None) -> str:
    if ev is None or not ev.items:
        return "No historical evidence was retrieved."
    top = ev.signals.top_similarity if ev.signals else ev.items[0].quality.semantic_relevance
    clusters = ", ".join(f"{c.action_class} x{c.support_count}" for c in ev.resolution_candidates[:3]) or "none"
    return (f"{len(ev.items)} similar historical cases retrieved; top similarity {top:.2f}; evidence {ev.sufficiency_level} ({ev.sufficiency_reason}); "
            f"resolution clusters: {clusters}; resolution confidence {ev.resolution_confidence:.2f}.")


def build_summary(st: AgentState) -> DecisionSummary:
    d, esc = st.decision, st.escalation
    action = d.action if d else "HUMAN_HANDOFF"
    reason_code = esc.reason_code if esc else "output_gate"
    rule = esc.rule if esc else "gate_default"
    policy_version = esc.policy_version if esc else ""
    blocking = list(d.blocking) if d else []
    ev_line = evidence_line(st.evidence)
    if action == "AUTO_HANDLE" and st.draft is not None and st.draft.strategy.value == "canned":
        return DecisionSummary(action=action, what_happened=f"Sent a fixed template ({rule}).", why=esc.reason if esc else "", evidence_basis="Templates make no product claims and need no evidence.",
                               next_step="No further action unless the customer replies.", reason_code=reason_code, rule=rule, policy_version=policy_version)
    if action == "AUTO_HANDLE":
        cov = st.verification.coverage if st.verification else 0.0
        basis = "; ".join(f"{c.evidence_id} ({c.action_class}, similarity {c.similarity:.2f}, rank {c.rank})" for c in st.citations) or "none"
        return DecisionSummary(action=action, what_happened=f"Answered automatically with a verified reply grounded in {len(st.citations)} historical case(s).",
                               why=f"{esc.reason if esc else ''} Evidence {st.evidence.sufficiency_level if st.evidence else ''}; verifier passed (evidence coverage {cov:.2f}); every output-gate check passed.",
                               evidence_basis=f"Cited cases: {basis}. {ev_line}", next_step="Send the reply; re-open the case if the customer reports the fix did not work.",
                               reason_code=reason_code, rule=rule, policy_version=policy_version)
    if action == "CLARIFICATION_REQUIRED":
        c = st.clarification
        asks = ", ".join(c.missing_information) if c and c.missing_information else "a description of the issue"
        return DecisionSummary(action=action, what_happened="Asked the customer one clarifying question instead of answering.",
                               why=(c.why if c else (esc.reason if esc else "")), evidence_basis=ev_line,
                               next_step=f"Wait for the customer's reply ({asks}); the same pipeline runs again with the added detail.",
                               reason_code=reason_code, rule=rule, policy_version=policy_version)
    why = esc.reason if esc else "Execution did not reach the policy stage."
    if esc is not None and esc.decision == "auto_handle" and blocking:
        why = f"The policy allowed automation but the output gate blocked the reply: {', '.join(blocking)}."
    next_step = st.handoff.recommended_next_action if st.handoff else "Review the thread and respond manually."
    return DecisionSummary(action="HUMAN_HANDOFF", what_happened="Handed the case to a human agent with a handoff packet; no autonomous answer was sent.", why=why,
                           evidence_basis=ev_line, next_step=next_step, reason_code=reason_code, rule=rule, policy_version=policy_version, blocking_checks=blocking)

File: frontend/lib/explain.ts
/**
 * "Why did ResolveAI do this?" Built only from structured decision data: the API's outcome, evidence gate, policy rule, risk
 * flags, verification and output gate (or, for a trace, the events it recorded). No model is ever asked to explain itself, and
 * nothing here decides anything.
 */
import type { Action, AgentTrace, ResolveResponse } from "@/lib/api/types";
import { humanize } from "@/lib/format";
import { activeRiskFlags } from "@/lib/handoff-summary";
import { ACTION_META, EVIDENCE_LEVEL_META, GATE_CHECK_LABELS, HARD_BLOCK_FLAGS, RISK_FLAG_LABELS, SUFFICIENCY_REASON_LABELS, evidenceLevel, reasonMeta } from "@/lib/labels";
import { eventData } from "@/lib/trace";

export const EXPLANATION_HEADINGS: Record<Action, string> = {
  AUTO_HANDLE: "Why this response was allowed",
  CLARIFICATION_REQUIRED: "Why the agent asked for more information",
  HUMAN_HANDOFF: "Why a human was required",
};

export interface Explanation {
  action: Action;
  heading: string;
  decision: string;
  reason: string;
  evidence: string;
  policy: string;
  risk: string;
  verification: string;
  /** What would have changed the decision, stated from the ordered policy rules and the output gate (never predicted). */
  whatWouldChange: string;
  nextAction: string;
  source: "result" | "trace";
}

/**
 * The policy's rules in the order `decide()` checks them (resolveai/policy/escalation.py POLICY_RULES, served by /agent/profile).
 * A console test asserts this list equals the served one, so the explanation cannot drift from the code.
 */
export const POLICY_RULE_ORDER = [
  "safety",
  "prompt_injection",
  "security",
  "legal",
  "abusive",
  "canned:greeting",
  "account_access",
  "payment_billing",
  "private_info",
  "hardware",
  "repeat_contact",
  "human_requested",
  "vague_hostile",
  "canned:other",
  "canned:acknowledgement",
  "canned:non_latin_script",
  "insufficient_context",
  "other_non_closure",
  "general_complaint_clarify",
  "taxonomy_gap",
  "low_confidence",
  "canned:non_english",
  "conflicting_evidence",
  "evidence_gate",
  "grounding_gate",
  "llm_fallback",
  "default_auto",
] as const;

const HANDOFF_CONDITIONS =
  "a hard risk flag (safety, prompt injection, security, legal or media, abuse), an account, billing, private-information, hardware or repeat-contact case, conflicting evidence, a draft that failed verification, or an unavailable model";

export function whatWouldChange(action: Action, rule: string | null | undefined): string {
  const idx = rule ? POLICY_RULE_ORDER.indexOf(rule as (typeof POLICY_RULE_ORDER)[number]) : -1;
  const position = idx >= 0 ? `Rule ${rule} is number ${idx + 1} of ${POLICY_RULE_ORDER.length} in the policy order` : null;
  if (action === "AUTO_HANDLE") {
    if (rule === "default_auto") {
      return `This reply reached the last policy rule: none of the ${POLICY_RULE_ORDER.length - 1} earlier rules matched. It would have been a human handoff on ${HANDOFF_CONDITIONS}; a clarification on an unstated issue, low intent confidence or a clarifiable evidence gap; and any failed output-gate check would have withheld it.`;
    }
    return `${position ?? "A fixed template"}. Any risk rule earlier in the order would have handed the case to a human instead; a template carries no troubleshooting content, so it needs no evidence.`;
  }
  if (action === "CLARIFICATION_REQUIRED") {
    return `${position ? `${position}. ` : ""}An automatic reply would have needed a stated issue, acceptable intent confidence and SUFFICIENT or STRONG evidence, then a verified draft. The same gap with a high-impact risk, several intents or a reason that cannot be clarified would have gone to a human.`;
  }
  if (position) {
    return `${position}; the rules after it were not evaluated. An automatic reply would have needed no earlier rule to match, SUFFICIENT or STRONG evidence, a verified draft and every output-gate check passing.`;
  }
  return "This handoff came from a failure or a gate outside the ordered policy rules (for example a dependency failure or an audit trail that could not be written); an automatic reply is never released in that case.";
}

const sufficiencyText = (reason: string | null | undefined) => SUFFICIENCY_REASON_LABELS[reason ?? ""] ?? humanize(reason);

function riskText(flags: string[], source?: string | null): string {
  if (!flags.length) return `No risk flags raised${source ? ` (${source})` : ""}.`;
  const hard = flags.filter((f) => HARD_BLOCK_FLAGS.has(f));
  const labels = flags.map((f) => RISK_FLAG_LABELS[f] ?? humanize(f));
  return `Raised: ${labels.join(", ")}${hard.length ? `. Hard block: ${hard.map((f) => RISK_FLAG_LABELS[f] ?? f).join(", ")}` : ""}${source ? ` (${source})` : ""}.`;
}

function gateText(gates: Record<string, boolean> | undefined, blocking: string[] | undefined): string {
  const checks = Object.keys(gates ?? {});
  if (!checks.length) return "";
  const blocked = blocking?.length ? blocking : checks.filter((k) => !gates![k]);
  return blocked.length
    ? ` Output gate: no automatic reply; ${blocked.length} of ${checks.length} checks not met (${blocked.map((k) => (GATE_CHECK_LABELS[k] ?? humanize(k)).toLowerCase()).join("; ")}).`
    : ` Output gate: all ${checks.length} checks passed.`;
}

export function explainResult(r: ResolveResponse): Explanation {
  const o = r.outcome;
  const level = evidenceLevel(r.evidence.sufficiency_level);
  const used = r.response.evidence_refs.length;
  const v = r.verification;
  let verification: string;
  if (v) {
    verification = v.verified
      ? "The drafted reply was verified against the evidence it cites."
      : `The drafted reply failed verification${v.issues.length ? `: ${v.issues.join("; ")}` : ""}. It was not sent.`;
  } else if (r.response.kind === "template_reply") {
    verification = "Not needed: a fixed template with no troubleshooting content.";
  } else {
    verification = "Not run: no reply was drafted, so there was nothing to verify.";
  }
  verification += gateText(o.output_gate, o.blocking_checks);
  return {
    action: r.action,
    heading: EXPLANATION_HEADINGS[r.action],
    source: "result",
    decision: `${ACTION_META[r.action].label}. ${o.what_happened}`,
    reason: r.action === "AUTO_HANDLE" ? o.why : `${reasonMeta(o.reason_code).label}. ${o.why}`,
    evidence:
      `${EVIDENCE_LEVEL_META[level].label} (${sufficiencyText(r.evidence.sufficiency_reason)}). ${r.evidence.items.length} historical case${r.evidence.items.length === 1 ? "" : "s"} retrieved, ${used} used in the response.` +
      (r.evidence.sufficient ? "" : " Not sufficient to answer safely; retrieved cases were not treated as an answer."),
    policy: `Rule ${o.rule} (reason ${o.reason_code}) in ${o.policy_version} was the first rule that matched. Autonomous response ${o.autonomous_response_allowed ? "allowed" : "not allowed"}.`,
    risk: riskText(activeRiskFlags(r.risk as unknown as Record<string, unknown>)),
    verification,
    whatWouldChange: whatWouldChange(r.action, o.rule),
    nextAction: o.next_step,
  };
}

/* ---------------------------------------------------------------- traces (no customer or evidence text is stored) */

interface PolicyEvent {
  decision?: string;
  reason_code?: string;
  rule?: string;
  policy_version?: string;
  clarification_allowed?: boolean;
  strategy?: string;
}
interface GateEvent {
  action?: string;
  blocking?: string[];
  gates?: Record<string, boolean>;
}
interface IntentEvent {
  intent?: string;
  confidence?: number;
  band?: string;
}
interface SecondOpinionEvent {
  before?: string;
  after?: string;
  applied?: boolean;
}
interface EvidenceEvent {
  sufficient?: boolean;
  reason?: string;
  level?: string;
  resolution_confidence?: number;
  consistency?: string;
  gate_version?: string;
}
interface RiskEvent {
  flags?: string[];
  source?: string;
}
interface ClarificationEvent {
  reason_code?: string;
  missing?: string[];
  already_provided?: string[];
}
interface RetrievalEvent {
  n_retrieved?: number;
  evidence_ids?: string[];
  top_similarity?: number;
  retriever?: string;
}

/** The decision data an audit trace records (no customer text, no evidence text). */
export function summarizeTraceDecision(trace: AgentTrace) {
  const d = <T,>(name: string) => eventData<T>(trace.events, name);
  const policy = d<PolicyEvent>("escalation_decided");
  const handoff = d<{ reason_code?: string }>("handoff_created");
  const clarification = d<ClarificationEvent>("clarification_created");
  const gate = d<GateEvent>("output_allowed") ?? d<GateEvent>("decision_made");
  const intent = d<IntentEvent>("intent_predicted");
  const second = d<SecondOpinionEvent>("second_opinion_used");
  const evidence = d<EvidenceEvent>("evidence_evaluated");
  const risk = d<RiskEvent>("risk_flags_extracted");
  const retrieval = d<RetrievalEvent>("retrieval_completed");
  const action = ((trace.final_decision ?? gate?.action) as Action | undefined) ?? null;
  const reasonCode = handoff?.reason_code ?? clarification?.reason_code ?? policy?.reason_code ?? "none";
  return { action, reasonCode, policy, gate, intent, second, evidence, risk, retrieval, clarification, finalIntent: second?.applied ? second.after : intent?.intent };
}

const TRACE_NEXT_ACTION: Record<Action, string> = {
  AUTO_HANDLE: "The verified reply was returned to the caller. In this reference environment nothing is sent to a customer.",
  CLARIFICATION_REQUIRED: "The clarifying question was returned; the conversation continues when the customer answers.",
  HUMAN_HANDOFF: "A human agent takes the case with the handoff packet. Only a holding notice is returned.",
};

export function explainTrace(trace: AgentTrace): Explanation | null {
  const s = summarizeTraceDecision(trace);
  if (!s.action) return null;
  const laterStage = Boolean(s.policy?.reason_code && s.policy.reason_code !== s.reasonCode);
  const stage = (k: string) => (trace.stage_status as Record<string, string> | undefined)?.[k];
  const verifyStatus = stage("verification");
  const level = evidenceLevel(s.evidence?.level);
  return {
    action: s.action,
    heading: EXPLANATION_HEADINGS[s.action],
    source: "trace",
    decision: `${ACTION_META[s.action].label}.`,
    reason:
      s.action === "AUTO_HANDLE"
        ? ACTION_META.AUTO_HANDLE.summary
        : `${reasonMeta(s.reasonCode).label}.${laterStage ? ` The policy stage chose ${humanize(s.policy?.reason_code).toLowerCase()}; a later stage changed the outcome.` : ""}`,
    evidence: s.evidence
      ? `${EVIDENCE_LEVEL_META[level].label} (${sufficiencyText(s.evidence.reason)}).${s.retrieval ? ` ${s.retrieval.n_retrieved ?? 0} candidates retrieved.` : ""} Evidence text is not stored in traces.`
      : "The evidence gate did not run.",
    policy: s.policy ? `Rule ${s.policy.rule ?? "n/a"} (reason ${s.policy.reason_code ?? "n/a"}) in ${s.policy.policy_version ?? trace.versions?.policy ?? "n/a"}.` : "No policy decision recorded.",
    risk: riskText(s.risk?.flags ?? [], s.risk?.source),
    verification: `${verifyStatus ? `Verification stage: ${humanize(verifyStatus).toLowerCase()}.` : "Verification stage not recorded."}${gateText(s.gate?.gates, s.gate?.blocking)}`,
    whatWouldChange: whatWouldChange(s.action, s.policy?.rule),
    nextAction: TRACE_NEXT_ACTION[s.action],
  };
}
